lib/runner.js

```diff
@@ -564,10 +564,16 @@ internals.protect = function (item, state, callback) {
     setImmediate(() => {
 
         domain.enter();
-        item.fn.call(null, (err) => {
+        const itemResult = item.fn.call(null, (err) => {
 
             finish(err, 'done');
         });
+        if (itemResult && itemResult.then instanceof Function) {
+            itemResult.then(() => finish(null), (err) => finish(err, 'done'));
```

Comment:
I thought you might mean that – but that would be incorrect. If the promise resolves with a value, `then(finish)` would pass the resolve value to `finish`, which would interpret it as an error!